Produce PDF files one page at a time if desired (BL-3929)

StephenMcConnel opened the pull request:
This allows low-memory machines (at least for Windows) to create
PDF files that otherwise would require too much memory.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/bloombooks/bloomdesktop/1311)
<!-- Reviewable:end -->

hatton commented:
Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/BloomExe/Publish/MakePdfUsingGeckofxHtmlToPdfProgram.cs, line 42 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1311#-KWd9anP09HXmTZ14U-U:-KWd9anP09HXmTZ14U-V:bd7ho5a) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/a954296f2723f825fb991bf0cdd86e8f864fae5f/src/BloomExe/Publish/MakePdfUsingGeckofxHtmlToPdfProgram.cs#L42)):*
> ```Smalltalk
> 	{
> 		public void MakePdf(string inputHtmlPath, string outputPdfPath, string paperSizeName,
> 			bool landscape, bool singlePageMode, Control owner, BackgroundWorker worker, DoWorkEventArgs doWorkEventArgs)
> ```

singlePageMode -->saveMemoryMode, everywhere

---

*[src/BloomExe/Publish/MakePdfUsingGeckofxHtmlToPdfProgram.cs, line 130 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1311#-KWdBDK3hyZWn4x1GtOQ:-KWdBDK3hyZWn4x1GtOR:bem97wh) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/a954296f2723f825fb991bf0cdd86e8f864fae5f/src/BloomExe/Publish/MakePdfUsingGeckofxHtmlToPdfProgram.cs#L130)):*
> ```Smalltalk
> 					@"Error message displayed in a message dialog box. {0} is the filename, {1} is a newline character.");
> 
> 				var msg2 = L10NSharp.LocalizationManager.GetDynamicString(@"Bloom", @"MakePDF.TrySinglePage",
> ```

Please add comment: this is message is intentionally separate because it was added after most languages were already localized.

---

*[src/BloomExe/Publish/MakePdfUsingGeckofxHtmlToPdfProgram.cs, line 131 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1311#-KWd9kz9S5SC8QHhqq4Z:-KWd9kz9S5SC8QHhqq4_:b-uik5tr) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/a954296f2723f825fb991bf0cdd86e8f864fae5f/src/BloomExe/Publish/MakePdfUsingGeckofxHtmlToPdfProgram.cs#L131)):*
> ```Smalltalk
> 
> 				var msg2 = L10NSharp.LocalizationManager.GetDynamicString(@"Bloom", @"MakePDF.TrySinglePage",
> 					"If the problem appears to be running out of memory, try checking the \"Create PDF File One Page At A Time\" item in the page size and orientation pulldown menu and try again.",
> ```

The book's images might have exceeded the amount of RAM memory available. Please turn on the \"Use Less Memory\" option which is slower but uses less memory.",

This does not really tell people where to find it, but the more descriptive one didn't either, I'm afraid, because it's doesn't have a label. So let's error on keeping it simple. This way also lets us move that control if needed without having to do a ton of re-localizing. We might for example make that menu say "Options".

---

*[src/BloomExe/Publish/MakePdfUsingGeckofxHtmlToPdfProgram.cs, line 172 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1311#-KWdBa0HLwdwbv45NLYz:-KWdBa0HLwdwbv45NLZ-:byp231g) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/a954296f2723f825fb991bf0cdd86e8f864fae5f/src/BloomExe/Publish/MakePdfUsingGeckofxHtmlToPdfProgram.cs#L172)):*
> ```Smalltalk
> 				bldr.Append(" -Landscape");
> 			if (singlePageMode)
> 				bldr.Append(" --single-pages");
> ```

This is a misleading argument name, if the program actually puts them back together before returning. If that's the case, it should just be something like "--economy-mode" or "--reduce-memory-use".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1311)*
<!-- Sent from Reviewable.io -->

hatton commented:
Review status: 0 of 5 files reviewed at latest revision, 6 unresolved discussions.

---

*[src/BloomExe/Publish/PublishView.cs, line 43 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1311#-KWdCKN40GuazmkJhI6z:-KWdCKN40GuazmkJhI7-:bcjj676) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/a954296f2723f825fb991bf0cdd86e8f864fae5f/src/BloomExe/Publish/PublishView.cs#L43)):*
> ```Smalltalk
> 		/// </summary>
> 		/// <value>The single page mode.</value>
> 		public bool SinglePageMode { get; set; }
> ```

LessMemoryMode

---

*[src/BloomExe/Publish/PublishView.cs, line 349 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1311#-KWdC54cUni_ADSruA_C:-KWdC54cUni_ADSruA_D:b1zh4ml) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/a954296f2723f825fb991bf0cdd86e8f864fae5f/src/BloomExe/Publish/PublishView.cs#L349)):*
> GetDynamicString("Bloom", "lessMemoryPdfMode", "Use less memory (slower)");

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1311)*
<!-- Sent from Reviewable.io -->

StephenMcConnel commented:
Review status: 0 of 5 files reviewed at latest revision, 6 unresolved discussions.

---

*[src/BloomExe/Publish/MakePdfUsingGeckofxHtmlToPdfProgram.cs, line 42 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1311#-KWd9anP09HXmTZ14U-U:-KWdkz50Lm5zIpltLN9F:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/a954296f2723f825fb991bf0cdd86e8f864fae5f/src/BloomExe/Publish/MakePdfUsingGeckofxHtmlToPdfProgram.cs#L42)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary>
> singlePageMode -->saveMemoryMode, everywhere
</details>
Done.

---

*[src/BloomExe/Publish/MakePdfUsingGeckofxHtmlToPdfProgram.cs, line 130 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1311#-KWdBDK3hyZWn4x1GtOQ:-KWdm2ConCtc1IjkJskY:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/a954296f2723f825fb991bf0cdd86e8f864fae5f/src/BloomExe/Publish/MakePdfUsingGeckofxHtmlToPdfProgram.cs#L130)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary>
> Please add comment: this is message is intentionally separate because it was added after most languages were already localized.
</details>
Done.

---

*[src/BloomExe/Publish/MakePdfUsingGeckofxHtmlToPdfProgram.cs, line 131 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1311#-KWd9kz9S5SC8QHhqq4Z:-KWdm1JjsbNgenole6g7:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/a954296f2723f825fb991bf0cdd86e8f864fae5f/src/BloomExe/Publish/MakePdfUsingGeckofxHtmlToPdfProgram.cs#L131)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary>
> The book's images might have exceeded the amount of RAM memory available. Please turn on the \"Use Less Memory\" option which is slower but uses less memory.",
> 
> This does not really tell people where to find it, but the more descriptive one didn't either, I'm afraid, because it's doesn't have a label. So let's error on keeping it simple. This way also lets us move that control if needed without having to do a ton of re-localizing. We might for example make that menu say "Options".
</details>
Done.

---

*[src/BloomExe/Publish/MakePdfUsingGeckofxHtmlToPdfProgram.cs, line 172 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1311#-KWdBa0HLwdwbv45NLYz:-KWdl0Tmw7LruQS8gzpd:br6aqor) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/a954296f2723f825fb991bf0cdd86e8f864fae5f/src/BloomExe/Publish/MakePdfUsingGeckofxHtmlToPdfProgram.cs#L172)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary>
> This is a misleading argument name, if the program actually puts them back together before returning. If that's the case, it should just be something like "--economy-mode" or "--reduce-memory-use".
</details>
Done.  I changed the GeckofxHtmlToPdf program to use --reduce-memory-use for the command line option.  (and renamed some its internal variables accordingly)

---

*[src/BloomExe/Publish/PublishView.cs, line 43 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1311#-KWdCKN40GuazmkJhI6z:-KWdlHDAd164YUikIDbC:b8jq1m7) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/a954296f2723f825fb991bf0cdd86e8f864fae5f/src/BloomExe/Publish/PublishView.cs#L43)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary>
> LessMemoryMode
</details>
Done.  Except I changed it to ReducedMemoryMode

---

*[src/BloomExe/Publish/PublishView.cs, line 349 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1311#-KWdC54cUni_ADSruA_C:-KWdnc6afXwcrLEKFD6S:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/a954296f2723f825fb991bf0cdd86e8f864fae5f/src/BloomExe/Publish/PublishView.cs#L349)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary>
> > GetDynamicString("Bloom", "lessMemoryPdfMode", "Use less memory (slower)");
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1311)*
<!-- Sent from Reviewable.io -->
